src/chainlink/oracle/ChainlinkOracleImpl.sol

```diff
@@ -0,0 +1,264 @@
+
+    error InvalidPair_FeedNotSet(QuotePair qp);
+    error StalePrice(AggregatorV3Interface feed, uint256 timestamp);
+    error NegativePrice(AggregatorV3Interface feed, int256 price);
+
+    /// -----------------------------------------------------------------------
+    /// structs
+    /// -----------------------------------------------------------------------
+
+    struct InitParams {
+        address owner;
+        bool paused;
+        SetPairDetailParams[] pairDetails;
+    }
+
+    struct SetPairDetailParams {
+        QuotePair quotePair;
+        PairDetail pairDetail;
+    }
+
+    struct PairDetail {
+        // encoded feed[]
+        bytes path;
+        bool inverted;
+    }
+
+    struct Feed {
+        AggregatorV3Interface feed;
+        /// @dev staleAfter > 1 hours
+        uint32 staleAfter;
+        /// @dev decimals should be same as feed.decimals()
+        uint8 decimals;
+        /// @dev mul should be true for the first feed in the path
+        bool mul;
+    }
+
+    /// -----------------------------------------------------------------------
+    /// events
+    /// -----------------------------------------------------------------------
+
+    event SetPairDetails(SetPairDetailParams[] params);
+
+    /// -----------------------------------------------------------------------
+    /// storage
+    /// -----------------------------------------------------------------------
+
+    /// -----------------------------------------------------------------------
+    /// storage - constants & immutables
+    /// -----------------------------------------------------------------------
+
+    address public immutable weth9;
+    address public immutable chainlinkOracleFactory;
+
+    /// -----------------------------------------------------------------------
+    /// storage - mutables
+    /// -----------------------------------------------------------------------
+
+    /// @notice The pair details
+    mapping(address => mapping(address => PairDetail)) internal $_pairDetails;
+
+    /// -----------------------------------------------------------------------
+    /// constructor & initializer
+    /// -----------------------------------------------------------------------
+
+    /**
+     * @notice constructor
+     * @param weth9_ The WETH9 contract address
+     */
+    constructor(address weth9_) {
+        weth9 = weth9_;
+        chainlinkOracleFactory = msg.sender;
+    }
+
+    /**
+     * @notice Initialize the contract
+     * @param params_ The init params
+     */
+    function initializer(InitParams calldata params_) external {
+        // only chainlinkOracleFactory may call `initializer`
+        if (msg.sender != chainlinkOracleFactory) revert Unauthorized();
+
+        __initOwnable(params_.owner);
+        $paused = params_.paused;
+
+        _setPairDetails(params_.pairDetails);
+        emit SetPairDetails(params_.pairDetails);
```

Comment:
i believe we didn't emit this in last oracle bc we relied on the factory's events for init

Reply:
i think we can get rid of this? lmk if you disagree